### ISSUE TITLE
ScErrStackOverflow from Typed JSON - am I modelling it wrong?

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3147]
**Foo.json** (plain JSON, no code behind)
```
{
  "Value": "",
  "Next": { "$" : { "Reuse":  "Foo"} }
}
```

**Program.cs**
```C#
Handle.GET("/foo", () => {
  return new Foo();
});
```

makes the code host crash with

> ScErrStackOverflow (SCERR1011): A stack overflow has been detected.

Im trying to model the equivalent of this:

```C#
class Foo {
  string Value;
  Foo Next;
}
```

**Comment by chrhol** [comment]
I'm a bit suprised that this pattern doesn't work, but thinking about it, it's quite obvious that this will happen.

I would say that there is nothing wrong in your code, the problem is how values for properties in TypedJson are handled internally. All values are initialized to a default value when a an instance of TypedJson is created, and for objects and arrays the default value is an empty instance instead of null.

Which leads to as soon you create a new `Foo`, the `Next` property will be initialized to a new instance of `Foo` and so on.

To fix this requires us to rethink how values are initialized and support objects and arrays to be null. Maybe we can implement some lazy initialization for objects to avoid breaking existing code that assumes they will never be null.

In the same time we should probably look at the other issues concerning null values in TypedJson, #269, #3072

**Note** as a workaround, if you really want to use this now, you can change the Next-property to be an array instead which will contain one (or zero) items pointing to next `Foo`.

```javascript
{
  "Value": "",
  "Next": [ { "$" : { "Reuse":  "Foo"} } ]
}
```

**Comment by per-samuelsson** [comment]
> [...] but thinking about it, it's quite obvious that this will happen.

Yeah, I came to the same conclusion. When it happened, and after I thought some about what could cause it, it was quite obvious. But I still wanted to open an issue, since we must find a way for this pattern and also make sure we don't crash the code host. :smile: 

Given that there are quite a few issues around Typed JSON, and the topic is quite broad and not always straightforward, we should try to put more resources on it. Feels like you are the only one right now @chrhol.

**Comment by chrhol** [comment]
> Given that there are quite a few issues around Typed JSON, and the topic is quite broad and not always straightforward, we should try to put more resources on it. Feels like you are the only one right now @chrhol.

Yes, I agree. There are more and more of these kind of issues where the question is more how we want it to work, or issues which require bigger parts of the code to be reworked.

**Comment by per-samuelsson** [comment]
> Yes, I agree. There are more and more of these kind of issues where the question is more how we want it to work, or issues which require bigger parts of the code to be reworked.

Exactly. I feel guilty everytime I report something, just because it feels like I put more on you. :santa: 

We should probably sit down a few of us at some point and discuss "Typed JSON 2.0", based on all the input we've collected.

**Comment by miyconst** [comment]
I have met the same exception for no visible reason. Restarting Starcounter server solved the issue. In the logs there is only one entry:

```
ScErrStackOverflow (SCERR1011): A stack overflow has been detected.
```

This is very frustrated when no additional details provided on where the error has occur and database just crashes.

**Comment by bigwad** [comment]
I'd suggest to disable all interception of exceptions and let the system generate a crash dump.
I don't know which exact code has prevented generating crash dump in your particular case, but I'm aware of two potential places:
- https://github.com/Starcounter/Starcounter/blob/develop/src/Hosting/sccoreapp/init.cpp#L24
- https://github.com/Starcounter/Starcounter/blob/develop/src/Hosting/Starcounter.Hosting/ExceptionManager.cs#L91

I have all this code removed in my local builds and it works perfectly.

**Comment by miyconst** [comment]
![image](https://cloud.githubusercontent.com/assets/6435556/18142599/a0f83502-6fbe-11e6-845f-72865f27c810.png)

@chrhol is it something you can investigate?
